packages/babel-plugin-transform-react-constant-elements/src/index.js

```diff
@@ -83,30 +146,70 @@ export default declare((api, options) => {
     visitor: {
       JSXElement(path) {
         if (HOISTED.has(path.node)) return;
-        HOISTED.add(path.node);
+        HOISTED.set(path.node, path.scope);
 
-        const state = { isImmutable: true };
+        const name = path.get("openingElement.name").node;
 
         // This transform takes the option `allowMutablePropsOnTags`, which is an array
         // of JSX tags to allow mutable props (such as objects, functions) on. Use sparingly
         // and only on tags you know will never modify their own props.
+        let mutablePropsAllowed = false;
         if (allowMutablePropsOnTags != null) {
           // Get the element's name. If it's a member expression, we use the last part of the path.
           // So the option ["FormattedMessage"] would match "Intl.FormattedMessage".
-          let namePath = path.get("openingElement.name");
-          while (namePath.isJSXMemberExpression()) {
-            namePath = namePath.get("property");
+          let lastSegment = name;
+          while (t.isJSXMemberExpression(lastSegment)) {
+            lastSegment = lastSegment.property;
           }
 
-          const elementName = namePath.node.name;
-          state.mutablePropsAllowed =
-            allowMutablePropsOnTags.indexOf(elementName) > -1;
+          const elementName = lastSegment.name;
+          mutablePropsAllowed = allowMutablePropsOnTags.includes(elementName);
         }
 
-        // Traverse all props passed to this element for immutability.
-        path.traverse(immutabilityVisitor, state);
+        // In order to avoid hoisting unnecessarily, we need to know which is
+        // the scope containing the current JSX element. If a parent of the
+        // current element has already been hoisted, we can consider its target
+        // scope as the base scope for the current element.
```

Comment:
Can you explain this bit? I can't really understand why we need to track scopes here, or in the analyzer, since it's traversing downwards.

In my head, we need to collect all of the `ReferencedIdentifier`s, then traverse upwards until we find a scope that defines one of them, and inject the lazy binding in that scope.

Reply:
<!--
In order to collect the bindings, we would need to traverse scopes in the analyzer anyway. This is because in this code

```js
function f() {
  let B;
  return () => {
    let A;
    return () => {
      return <X a={() => { let A; return A }} b={() => { return B }} />
    }
  }
}
```

we cannot just collect all the `ReferencedIdentifiers` but we need to differentiate between `A` (declared inside) and `B` (declared outside).
-->

When you have this code:
```js
function fn() {
  return <div><p/></div>;
}
```

first we traverse the `div`, and it becomes
```js
var _div;

function fn() {
  return _div || (_div = <div><p/></div>);
}
```

and then we don't want it to become
```js
var _div, _p;

function fn() {
  return _div || (_div = <div>{_p || (_p = <p/>)}</div>);
}
```
because it's unnecessary.

However, we might still want to hoist `<p/>` in some cases:
```js
function fn() {
  let cl = "a";
  return () => <div className={cl}><p/></div>;
}
```
will first become
```js
function fn() {
  var _div;
  let cl = "a";
  return () => _div || (_div = <div className={cl}><p/></div>);
}
```
and then it can be still optimized to
```js
var _p;
function fn() {
  var _div;
  let cl = "a";
  return () => _div || (_div = <div className={cl}>{_p || (_p = <p/>)}</div>);
}
```